### ISSUE TITLE
tuple scopes should nest

### ISSUE THREAD

**Issue by kurtbrose** [mahmoud/glom#95]
just messing around, ran into this pattern:

```python
    ('Rules', ([(
        Assign(S['exp'], glom.Coalesce('Expiration.Days', default=None)),
        'Filter',
        Coalesce(('And.Tags', ['Value']), ('Tag.Value', lambda x: [x])),
        Coalesce((Check(validate=lambda t: lifespan in t), S['exp']), default=SKIP),
        )], T[0])
    )
```

basically, using `Assign(S)` would be a handy way to stash something in the scope to be fetched later

BUT this requires that tuples nest their scopes so that the later sections of the tuple can see earlier ones (this seems to be intuitively correct behavior since tuples correspond to sequential execution and later statements can see the scope created by earlier statements in a sequence)

**Comment by kurtbrose** [comment]
https://github.com/mahmoud/glom/pull/97 -- also some `T.__setitem__()` stuff in there that didn't work out

**Comment by mahmoud** [comment]
Agreed, on the intuitive behavior, now to review the PR.

**Comment by mahmoud** [comment]
Shipped in 19.10.0!
